#46 Enhances the functional call grammar to accept variables/identifi…

iwarapter opened the pull request:
…ers.

racodond commented on puppet-squid/src/main/java/com/iadams/sonarqube/puppet/api/PuppetGrammar.java:
```diff
@@ -162,7 +162,10 @@ public static void grammar(LexerfulGrammarBuilder b) {
-        b.rule(ARGUMENT_EXPRESSION_LIST).is(EXPRESSION, b.zeroOrMore(COMMA, EXPRESSION), b.optional(COMMA));
+        b.rule(ARGUMENT_EXPRESSION_LIST).is(
+                b.firstOf(EXPRESSION, VARIABLE, QUALIFIED_IDENTIFIER, IDENTIFIER),
+                b.zeroOrMore(COMMA, EXPRESSION),
```
@iwarapter: The grammar is getting quite complex... There's something wrong here. Instead of `EXPRESSION`, there should be `EXPRESSION, VARIABLE, QUALIFIED_IDENTIFIER, IDENTIFIER`. It parses for now because all  the arguments after the first one are wrongly considered as function calls even if they're not.

![image](https://cloud.githubusercontent.com/assets/1215713/8796032/f2c6ba98-2f91-11e5-9b8c-484b96335059.png)

iwarapter commented:
@racodond i just came across this https://github.com/puppetlabs/puppet-specifications/blob/master/language/lexical_structure.md so im going to review pretty much everything. How familiar are you with sonar sslr grammar stuff (you may have guessed this is my first attempt with it)

racodond commented:
@iwarapter: Indeed, this file as well as the others one level up (such as https://github.com/puppetlabs/puppet-specifications/blob/master/language/catalog_expressions.md) should be quite nice to rewrite the grammar. It's gonna be a fairly big work though :-)

I'm not familiar with grammar and SSLR grammar. I wrote a fairly simple one a few days ago to provide a Java Properties plugin: https://github.com/racodond/sonar-jproperties-plugin/blob/master/jproperties-squid/src/main/java/org/sonar/jproperties/parser/JavaPropertiesGrammar.java. I also fixed the existing CSS grammar while writing new rules: https://github.com/SonarCommunity/sonar-css/blob/master/css-squid/src/main/java/org/sonar/css/parser/CssGrammar.java. But I'm really a newbie.

I would try to mimic what has been done for JavaScript: https://github.com/SonarCommunity/sonar-javascript/tree/master/javascript-squid/src/main/java/org/sonar/javascript. What do you think?

And I'd be glad to dive into grammar stuff with you.

iwarapter commented:
Yeah it is going to be a big bit of work but i think its necessary otherwise we will end up re-writing all the rules later i'd rather focus on this now. 

I originally based this plugin on the python plugin https://github.com/SonarCommunity/sonar-python so you will see many similarities between the two.

I'm trying to think of the best way to break down the task, i'd like to clean up the lexer to begin with as some might not quite line up and i'd like to support single/double quoted strings, and this will give us more freedom on the rules later on.

So after that its how we break down the task. I think the core expressions are OK (although now all are currently implemented). One of the key things is how we handle the data types as i have varying degrees of success mapping this but as we have seen on some of the errors its abit hit and miss.

So id suggest looking at things in this order?
- Lexer
- Expressions (review)
- Data Types
- Complex grammar definitions

Thoughts?

racodond commented:
I agree that the grammar should be rewritten. Your plan sounds good to me. I guess that it's gonna be hard to both work on it at the same time.  So, I'll let you work on it. When you have something pretty solid, I'm ready to run extensive tests and fine tune the grammar.

iwarapter commented:
Nice i cleaned up the lexer a little, broke about half the tests :+1:, this is going to be alot of refactoring.

racodond commented:
Go go go :-)

David RACODON
Freelance QA Consultant
LinkedIn https://ch.linkedin.com/pub/david-racodon/11/62/283 | Twitter
https://twitter.com/davidracodon

On Wed, Jul 22, 2015 at 11:28 PM, iwarapter notifications@github.com
wrote:

> Nice i cleaned up the lexer a little, broke about half the tests [image:
> :+1:], this is going to be alot of refactoring.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/iwarapter/sonar-puppet/pull/55#issuecomment-123871977
> .

iwarapter commented:
@racodond which version of puppet are you using? We are currently dealing with issues across versions 3.x and 4.x of puppet.

Since puppet enterprise is currently 3.8 i'd like to focus on that. Once thats stable we can move onto 4.x.

racodond commented:
We're running Puppet Enterprise 3.8. I agree with you; we can focus on
version 3.8.

racodond commented:
How's the rewrite going?

iwarapter commented:
Coming along, have improved the lexer and was just fixing all the tests it broke, so i can improve the grammar and fix all the tests that breaks. I can publish what i have on another branch if your keen to see?

racodond commented:
Sure. Tell me if we can split work and if I can help you in any way.

iwarapter commented:
I have pushed the code to a new branch (major_grammar_rework) i have updated the lexer to identify single/double quoted strings, and cleaned up how it identifies names and references (https://github.com/puppetlabs/puppet-specifications/blob/master/language/lexical_structure.md#identifiers) I have been aligning things with the 3.8 definitions but having finished everything (spent alot of time fixing all the tests when i gutted the lexer) 

Testing and sanity checking of the grammar currently is best place to focus.

iwarapter commented:
One of the key mistakes i have realised i made was when defining things like resource references (http://docs.puppetlabs.com/puppet/3.8/reference/lang_datatypes.html#resource-references) i rather lazily did: 
b.rule(RESOURCE_REF).is(REF, ARRAY);
When i should of honoured the proper definition:
b.rule(RESOURCE_REF).is(REF, LBRACK, TITLE, b.zeroOrMore(COMMA, TITLE), RBRACK);

Spotting and fixing stuff like that is my current focus as that is what caused issues like above (where everything evaluated to a func_call)
